fdbclient/NativeAPI.actor.cpp

```diff
@@ -3509,3 +3510,91 @@ ACTOR Future<Void> snapCreate(Database inputCx, StringRef snapCmd, UID snapUID)
 	TraceEvent("SnapCreateComplete").detail("UID", snapUID);
 	return Void();
 }
+
+ACTOR Future<Void> snapshotDatabase(DatabaseContext* cx, StringRef snapPayload, UID snapUID, Optional<UID> debugID) {
+	TraceEvent("NativeAPI.SnapshotDatabaseEnter")
+		.detail("SnapPayload", snapPayload)
+		.detail("SnapUID", snapUID);
+	try {
+		if (debugID.present()) {
+			g_traceBatch.addEvent("TransactionDebug", debugID.get().first(), "NativeAPI.snapshotDatabase.Before");
+		}
+
+		ProxySnapRequest req(snapPayload, snapUID, debugID);
+		wait(loadBalance(cx->getMasterProxies(false), &MasterProxyInterface::proxySnapReq, req, cx->taskID, true /*atmostOnce*/ ));
+		if (debugID.present())
+			g_traceBatch.addEvent("TransactionDebug", debugID.get().first(),
+									"NativeAPI.SnapshotDatabase.After");
+	} catch (Error& e) {
+		TraceEvent("NativeAPI.SnapshotDatabaseError")
+			.detail("SnapPayload", snapPayload)
+			.detail("SnapUID", snapUID)
+			.error(e, true /* includeCancelled */);
+		throw;
+	}
+	return Void();
+}
+
+ACTOR Future<Void> snapCreateVersion2(Database inputCx, StringRef snapCmd, UID snapUID) {
+	state DatabaseContext* cx = inputCx.getPtr();
+	// remember the client ID before the snap operation
+	state UID preSnapClientUID = cx->clientInfo->get().id;
+
+	TraceEvent("SnapCreateEnterVersion2")
+	    .detail("SnapCmd", snapCmd.toString())
+	    .detail("UID", snapUID)
+	    .detail("PreSnapClientUID", preSnapClientUID);
+
+	StringRef snapCmdArgs = snapCmd;
+	StringRef snapCmdPart = snapCmdArgs.eat(":");
+	Standalone<StringRef> snapUIDRef(snapUID.toString());
+	Standalone<StringRef> snapPayloadRef = snapCmdPart
+		.withSuffix(LiteralStringRef(":uid="))
+		.withSuffix(snapUIDRef)
+		.withSuffix(LiteralStringRef(","))
+		.withSuffix(snapCmdArgs);
+
+	try {
+		Future<Void> exec = snapshotDatabase(cx, snapPayloadRef, snapUID, snapUID);
+		double timeOut = g_network->isSimulated() ? 80.0 : CLIENT_KNOBS->SNAP_CREATE_TIMEOUT;
+		wait(timeoutError(exec, timeOut));
+	} catch (Error& e) {
+		TraceEvent("SnapshotDatabaseErrorVersion2")
+			.detail("SnapCmd", snapCmd.toString())
+			.detail("UID", snapUID)
+			.error(e);
+		throw;
+	}
+
+	UID postSnapClientUID = cx->clientInfo->get().id;
+	if (preSnapClientUID != postSnapClientUID) {
+		// if the client IDs changed then we fail the snapshot
+		TraceEvent("UIDMismatchVersion2")
+		    .detail("SnapPreSnapClientUID", preSnapClientUID)
+		    .detail("SnapPostSnapClientUID", postSnapClientUID);
+		throw coordinators_changed();
+	}
+
+	TraceEvent("SnapCreateExitVersion2")
+	    .detail("SnapCmd", snapCmd.toString())
+	    .detail("UID", snapUID)
+	    .detail("PreSnapClientUID", preSnapClientUID);
+	return Void();
+}
+
+ACTOR Future<Void> snapCreate(Database cx, StringRef snapCmd, UID snapUID, int version) {
+	if (version == 1) {
+		wait(snapCreateVersion1(cx, snapCmd, snapUID));
+		return Void();
+	}
+	state int oldMode = wait( setDDMode( cx, 0 ) );
```

Comment:
Yes, we (Senthil and I) discussed this. He agreed that this code should be executed in the data distributor and it should be ephemeral (basically DD will be disabled but the state won't be written to the database - so if the DD fails, it will recover with DD enabled).

Reply:
This change will be done in a separate PR.